web3_proxy/src/stats/mod.rs

```diff
@@ -372,59 +377,22 @@ impl BufferedRpcQueryStats {
         // TODO: rename to owner_id?
         let sender_user_id = key.rpc_key_user_id.map_or(0, |x| x.get());
 
-        // Gathering cache and database rows
-        let user_balance = self
-            ._get_user_balance(sender_user_id, user_balance_cache, db_conn)
-            .await?;
-
-        let mut user_balance = user_balance.write().await;
-
-        let premium_before = user_balance.active_premium();
-
-        // First of all, save the statistics to the database:
-        let paid_credits_used = self
-            ._save_db_stats(chain_id, db_conn, &key, premium_before)
-            .await?;
-
-        // No need to continue if no credits were used
-        if self.sum_credits_used == 0.into() {
-            // write-lock is released
-            return Ok(());
-        }
-
-        // Update and possible invalidate rpc caches if necessary (if there was a downgrade)
-        {
-            user_balance.total_spent_paid_credits += paid_credits_used;
-
-            // Invalidate caches if remaining is getting close to $0
-            // It will be re-fetched again if needed
-            if premium_before && user_balance.remaining() < Decimal::from(1) {
-                let rpc_keys = rpc_key::Entity::find()
-                    .filter(rpc_key::Column::UserId.eq(sender_user_id))
-                    .all(db_conn)
-                    .await?;
-
-                // clear all keys owned by this user from the cache
-                for rpc_key_entity in rpc_keys {
-                    rpc_secret_key_cache
-                        .invalidate(&rpc_key_entity.secret_key.into())
-                        .await;
-                }
-            }
-        }
+        // save the statistics to the database:
+        self._save_db_stats(chain_id, db_conn, &key).await?;
 
-        if premium_before {
+        // Apply all the referral logic; let's keep it simple and flat for now
+        if self.paid_credits_used > 0.into() {
             // Start a transaction
             let txn = db_conn.begin().await?;
 
-            // Apply all the referral logic; let's keep it simple and flat for now
             // Calculate if we are above the usage threshold, and apply a bonus
             // Optimally we would read this from the balance, but if we do it like this, we only have to lock a single table (much safer w.r.t. deadlocks)
             // referral_entity.credits_applied_for_referrer * (Decimal::from(10) checks (atomically using this table only), whether the user has brought in >$100 to the referer
             // In this case, the sender receives $100 as a bonus / gift
             // Apply a 10$ bonus onto the user, if the user has spent 100$
             // TODO: i think we do want a LockType::Update on this
             match referee::Entity::find()
+                .lock(LockType::Update)
```

Comment:
why is this lock needed? we only do upserts, no?

Reply:
This one is fine:
```
                            referral_entity.one_time_bonus_applied_for_referee =
                                sea_orm::Set(bonus_for_user);
```

This one is not. This has a race:
```
                            referral_entity.credits_applied_for_referrer = sea_orm::Set(
                                referral_entity.credits_applied_for_referrer.as_ref()
                                    + referrer_bonus,
                            );
```